### DISCUSSION TITLE
Different behavior between  mg_http_reply() and mg_printf()

### DISCUSSION THREAD

**Discussion by MalcolmGillott** [cesanta/mongoose#2530] [Ask questions, get answers] [answered]
I am testing some code and would like to understand the difference in behavior in these two calls.
I have created some REST API code, and in one scenario, I am handling an invalid request to my API

If I use  
    mg_http_reply(c, 404, "", "", "")  
to respond to a bad request, then my API behaves how I expect

If I use
    mg_printf(c,"HTTP/1.1 404 Not Found\r\nContent-Length: 0            \r\n\r\n");
which generates the same reply, Mongoose will not fire another MG_EV_HTTP_MSG event until the connection is closed and re-opened.

Plainly I can use mg_http_reply() as my solution, but I would like to understand what is going on

Thanks
Malc

**Comment by scaprile** [comment, marked as answer]
**Please read https://mongoose.ws/documentation/#best-practices**

Long story short: use `mg_http_reply()` or chunked encoding (`mg_http_...chunk()`). Those functions set `Content-Length` and `c->is_resp`for you. If you use `mg_printf()` or `mg_send()` instead, don't forget to set `Content-Length` and `c->is_resp` yourself

Please follow the guidelines in the [documentation](https://mongoose.ws/documentation/),  [examples](https://github.com/cesanta/mongoose/tree/master/examples/) and [tutorials](https://mongoose.ws/documentation/#tutorials-and-examples) available.
